connectors/filtering/basic_rule.py

```diff
@@ -0,0 +1,167 @@
+        elif isinstance(parsed_date_or_datetime, datetime.date):
+            # adds 00:00 to the date and returns datetime
+            return datetime.datetime.combine(parsed_date_or_datetime, datetime.time.min)
+        else:
+            return value
+
+    except ParserError:
+        return value
+
+
+def to_bool(value):
+    if len(value) == 0 or re.match("^(false|f|no|n|off)$", value):
+        return False
+
+    if re.match("^(true|t|yes|y|on)$", value):
+        return True
+
+    return value
+
+
+def try_coerce(value):
+    coerced_value = to_float(value)
+
+    if isinstance(coerced_value, str):
+        coerced_value = to_datetime(value)
+
+    if isinstance(coerced_value, str):
+        coerced_value = to_bool(value)
+
+    return coerced_value
+
+
+class Rule(Enum):
+    EQUALS = 1
+    STARTS_WITH = 2
+    ENDS_WITH = 3
+    CONTAINS = 4
+    REGEX = 5
+    GREATER_THAN = 6
+    LESS_THAN = 7
+
+
+class Policy(Enum):
+    INCLUDE = 1
+    EXCLUDE = 2
+
+
+class BasicRule:
+    DEFAULT_RULE_ID = "DEFAULT"
+
+    def __init__(self, id_, order, policy, field, rule, value):
+        self._id = id_
+        self._order = order
+        self._policy = policy
+        self._field = field
+        self._rule = rule
+        self._value = value
+
+    @classmethod
+    def default_rule(cls):
+        return cls(
+            id_=BasicRule.DEFAULT_RULE_ID,
+            order=0,
+            policy=Policy.INCLUDE,
+            field="_",
+            rule=Rule.EQUALS,
+            value=".*",
+        )
+
+    def matches(self, document):
+        if self.is_default_rule():
+            return True
+
+        if self.field() not in document:
+            return False
+
+        document_value = document[self.field()]
+        coerced_rule_value = self.coerce_rule_value_based_on_document_value(
+            document_value
+        )
+
+        match self.rule():
+            case Rule.STARTS_WITH:
+                return str(document_value).startswith(self.value())
+            case Rule.ENDS_WITH:
+                return str(document_value).endswith(self.value())
+            case Rule.CONTAINS:
+                return self.value() in str(document_value)
+            case Rule.REGEX:
+                return re.match(self.value(), str(document_value)) is not None
+            case Rule.LESS_THAN:
+                return document_value < coerced_rule_value
+            case Rule.GREATER_THAN:
+                return document_value > coerced_rule_value
+            case Rule.EQUALS:
+                return document_value == coerced_rule_value
+
+    def id_(self):
+        return self._id
+
+    def order(self):
+        return self._order
+
+    def policy(self):
+        return self._policy
+
+    def field(self):
+        return self._field
+
+    def rule(self):
+        return self._rule
+
+    def value(self):
+        return self._value
+
+    def is_default_rule(self):
+        return self.id_() == BasicRule.DEFAULT_RULE_ID
```

Comment:
in Python each indirection slows down execution, so I would remove the `id_` function and use the attribute directly here